<h1>{{ page.title }}</h1>
<p class="meta">{{ page.date | date_to_string }}</p>

<div class="post">
  {{ content }}
</div>


<h2><u>References</u></h2>
1. LDSC: Bulik-Sullivan, B. K. et al. (2015) LD Score regression distinguishes confounding from polygenicity in genome-wide association studies. Nature Genetics, 47(3), 291–295.
2. Genetic Correlation: Bulik-Sullivan, B. K. et al. (2015) An atlas of genetic correlations across human diseases and traits. Nature Genetics, 47, 1236–1241.

<h2><u>1. How is Genetic Correlation Estimated?</u></h2>
The relationship between the product of the z-scores from two studies of traits with some genetic correlation is similar to that of a single polygenic trait: for a polygenic trait, <bold>SNPs with higher LD</bold> will have higher <bold>Chi-squared statistics</bold>, and SNPs with a high LD will also have <bold>higher z-score product<bold> across correlated traits. In fact, the method to obtain a genetic correlation estimate is analogous to the method used to estimate SNP-heritability:


```{r, echo=FALSE, out.width="80%", fig.align="center"}
knitr::include_graphics("/Users/alicesmail/Desktop/KCL/LCV/Images/LD-10.png")
```

Where the slope estimates genetic covariance per SNP (instead of heritability):

```{r, echo=FALSE, out.width="90%", fig.align="center"}
knitr::include_graphics("/Users/alicesmail/Desktop/KCL/LCV/Images/LD-11.png")
```

From this slope estimate, we can get the genetic correlation value using the equation:

```{r, echo=FALSE, out.width="35%", fig.align="center"}
knitr::include_graphics("/Users/alicesmail/Desktop/KCL/LCV/Images/LD-12.png")
```

Essentially, similar z-scores with a large magnitude for the same SNP in two different GWASes, will result in a larger the z-score product and a steeper regression line:

```{r, echo=FALSE, out.width="35%", fig.align="center"}
knitr::include_graphics("/Users/alicesmail/Desktop/KCL/LCV/Images/LD-15.png")
```

<h2><u>2. Estimating Genetic Correlation for BMI & T2D</u></h2>
We might hypothesize that the strong association of <bold>body mass index (BMI)</bold> with <bold>type 2 diabetes (T2D)</bold> might mean that these two traits have some level of genetic correlation. We can test this by performing LD score regression using the <bold>LDSC package</bold>, which estimates the <bold>heritability</bold> of single traits, as well as the <bold>genetic correlation</bold> between two traits.

LDSC software is employed through the command line, or via a bash script. The first step of our workflow is downloading our data from GWAS catalog:

```{bash}
# Define working directory
#WORKING=/UsingLDSC/BMI_T2D/

# Download GWAS summary statistics: T2D and BMI
#wget http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90029001-GCST90030000/GCST90029024/GCST90029024_buildGRCh37.tsv -P $WORKING
#wget http://ftp.ebi.ac.uk//pub/databases/gwas/summary_statistics/GCST90428001-GCST90429000/GCST90428119/GCST90428119.tsv.gz -P $WORKING
```

<bold><u>2a. Removing the MHC Region</u></h2>
Before we use LDSC, we can perform a prior step that removes the <bold>MHC region</bold>: the complex LD patterns in this region can affect heritability & genetic correlation estimates. Using the script RemoveMHC.py (at github.com/alicesmail12/LDSC) we can remove the MHC region from both of our downloaded files:

```{bash}
# Remove MHC region
#$WORKING/RemoveMHC.py \
#--sumstatsFile $WORKING/GWASFiles/GCST90428119.tsv \
#--sep t \
#--logFile $WORKING/Logs/BMIMHC \
#--outFile $WORKING/GWASFiles/BMIMHC            

#$WORKING/RemoveMHC.py \
#--sumstatsFile $WORKING/GWASFiles/GCST90029024_buildGRCh37.tsv \
#--sep t \
#--logFile $WORKING/Logs/T2DMHClog \
#--outFile $WORKING/GWASFiles/T2DMHC
```

We can inspect the output logs from these commands to check nothing went wrong:

```{bash}
cat /Users/alicesmail/Desktop/KCL/LCV/Logs/BMIMHCrm.log
```

<bold><u>2b. Preparing the LDSC Software</u></h2>
Now that the MHC region is removed from both GWAS files, we can set up an environment to perform LDSC. First we load Python 2.7, as this version works with LDSC:

```{bash}
# Within working dir activate python 2.7 (required for LDSC)
#cd $WORKING
#CONDA_SUBDIR=osx-64 conda create -n py27 python=2.7                               
#conda activate py27
#conda config --env --set subdir osx-64         
```

Then we can clone the LDSC software from the LDSC repository (github.com/bulik/ldsc):

```{bash}
# Clone LSDC software into working dir
#cd $WORKING
#git clone https://github.com/bulik/ldsc.git    
```

Once LDSC is cloned we can activate it and check it is working:

```{bash}
# Define software space
#SOFTWARE==/UsingLDSC/BMI_T2D/ldsc

# Activate LDSC
#cd ldsc   
#conda env create --file environment.yml
#conda activate ldsc

# Check LDSC is working
#./munge_sumstats.py -h
```

<bold><u>2c. Munging GWAS Files</u></h2>

If we can view the help information for munge_sumstats.py, LDSC has been successfully activated and we can use it on our prepared GWAS files. The first step is 'munging' both files, which <bold>standardises</bold> the format of each file and performs a <bold>psuedo-QC</bold> using a list of SNPs from HapMap3 (w_hm3.snplist). There are lots of flags you can specify to munge_sumstats.py - the ones I have specified below include:
<ul>
<li><bold>--sumstats<bold>: the GWAS file location</li>
<li><bold>--N</bold>: the GWAS sample size</li>
<li><bold>--snp<bold>: the name of the SNP column (that contains rsIDs)</li>
<li><bold>--merge-alleles<bold>: HapMap3 SNP file location</li>
<li><bold>--out<bold>: output file location</li>
</ul>

```{bash}
# Munge T2D GWAS
#$SOFTWARE/munge_sumstats.py \
#--sumstats $WORKING/GWASFiles/T2DMHC.tsv \
#--N 468298 \
#--snp SNP \
#--merge-alleles $WORKING/GWASFiles/w_hm3.snplist \
#--out $WORKING/T2D

# Munge BMI GWAS
#$SOFTWARE/munge_sumstats.py \
#--sumstats $WORKING/GWASFiles/BMIMHC.tsv \
#--N 342566 \
#--snp SNP \
#--merge-alleles $WORKING/GWASFiles/w_hm3.snplist \
#--out $WORKING/BMI
```

To check our files were munged without any issues, we can check the log:

```{bash}
cat /Users/alicesmail/Desktop/KCL/LCV/Logs/BMI.log
```

Lots of information is contained in the log output for BMI. For example, we can see:
<ul>
<li>There are <bold>5312</bold> SNPs significantly associated with BMI.</li>
<li><bold>16,799,176 SNPs</bold> were removed because they are not in the HapMap3 SNP list.</li>
<li><bold>885,937 SNPs</bold> were removed because they had missing values.</li>
<li>The median beta value was near 0 <bold>(1.96e-05)</bold>, which is expected.</li>
<li>The output file contains <bold>846,100 SNPs</bold> with beta values.</li>
</ul>

<bold><u>2d. Performing Genetic Correlation Analysis</u></h2>
If we are happy that our data has been correctly munged, we can also use LDSC to perform genetic correlation analysis:

```{bash}
# Get genetic correlation between BMI & T2D
#$SOFTWARE/ldsc.py \
#--rg $WORKING/GWASFiles/BMI.sumstats.gz,$WORKING/GWASFiles/T2D.sumstats.gz \
#--ref-ld-chr $WORKING/eur_w_ld_chr/ \
#--w-ld-chr $WORKING/eur_w_ld_chr/ \
#--out $WORKING/BMI_T2D 
```

Here we use some new flags:
<ul>
<li><bold>--rg</bold>: specifies the two munged GWAS files</li>
<li><bold>--ref-ld-chr</bold> & <bold>--w-ld-chr</bold> specify LD score files for separate chromosomes (eur_w_ld_chr)</li>
</ul>
The log output for this function gives contains our results:

```{bash}
cat /Users/alicesmail/Desktop/KCL/LCV/Logs/BMI_T2D.log
```

From the log output we can see:
<ul>
<li><bold>816475 SNPs</bold> were used for the genetic correlation analysis.</li>
<li>The <bold>heritability</bold> of BMI is <bold>0.25</bold>, while the heritability of T2D is <bold>0.04</bold>.</li>
<li>The genetic correlation of BMI & T2D is <bold>0.56</bold>, and is highly significant (<bold>p=1.74e-57</bold>).</li>
</ul>

<h2><u>3. Estimating Genetic Correlation for Tea Drinking & Height</u></h2>
We can repeat our workflow for two traits that we suspect might not be correlated, in order to see how our results differ:

```{bash}
#!/bin/bash

# Define working spaces
#WORKING=/UsingLDSC/Tea_Height/

# Download GWAS summary statistics: Tea and Height
#wget http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90096001-GCST90097000/GCST90096926/GCST90096926_buildGRCh37.tsv.gz -P $WORKING/GWASFiles
#wget http://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90267001-GCST90268000/GCST90267284/GCST90267284_buildGRCh37.tsv -P $WORKING/GWASFiles

# Remove MHC region
#$WORKING/RemoveMHC.py \
#--sumstatsFile $WORKING/GWASFiles/GCST90096926_buildGRCh37.tsv \
#--sep t \
#--logFile $WORKING/Logs/TeaMHClog \
#--outFile $WORKING/GWASFiles/TeaMHC            

#$WORKING/RemoveMHC.py \
#--sumstatsFile $WORKING/GWASFiles/GCST90267284_buildGRCh37.tsv \
#--sep t \
#--logFile $WORKING/Logs/HeightMHClog \
#--outFile $WORKING/GWASFiles/HeightMHC

# Within working dir activate python 2.7 (required for LDSC)
#cd $WORKING
#CONDA_SUBDIR=osx-64 conda create -n py27 python=2.7                               
#conda activate py27
#conda config --env --set subdir osx-64         

# Define software space
#SOFTWARE=/Users/alicesmail/Desktop/KCL/LCV/ldsc

# Activate LDSC
#cd ldsc   
#conda env create --file environment.yml
#conda activate ldsc

# Check LDSC is working
#./munge_sumstats.py -h

# Munge T2D GWAS
#$SOFTWARE/munge_sumstats.py \
#--out $WORKING/TeaMHC \
#--merge-alleles $WORKING/GWASFiles/w_hm3.snplist \
#--N 434171 \
#--snp SNP \
#--sumstats $WORKING/GWASFiles/TeaMHC.tsv

# Munge BMI GWAS
#$SOFTWARE/munge_sumstats.py \
#--out $WORKING/HeightMHC \
#--merge-alleles $WORKING/GWASFiles/w_hm3.snplist \
#--N 283749 \
#--snp SNP \
#--sumstats $WORKING/GWASFiles/HeightMHC.tsv

# Get genetic correlation between BMI & T2D
#$SOFTWARE/ldsc.py \
#--ref-ld-chr $WORKING/eur_w_ld_chr/ \
#--out $WORKING/Tea_Height \
#--rg $WORKING/GWASFiles/TeaMHC.sumstats.gz,$WORKING/GWASFiles/HeightMHC.sumstats.gz \
#--w-ld-chr $WORKING/eur_w_ld_chr/ 
```

We can inspect the log output:

```{bash}
cat /Users/alicesmail/Desktop/KCL/LCV/Logs/Tea_Height.log
```

Here we can see that:
<ul>
<li><bold>1021233 SNPs</bold> were used for the genetic correlation analysis.</li>
<li>The <bold>heritability</bold> of tea consumption is <bold>0.05</bold>, while the heritability of height is <bold>0.45</bold>.</li>
<li>The genetic correlation between tea consumption & height is <bold>0.03</bold>, which is not significant (<bold>p=0.12</bold>).</li>
</ul>

<h2><u>4. Conclusion</u></h2>
Here we have performed genetic correlation analysis on two pairs of traits using the LDSC package. We can see that while one pair of traits (T2D & BMI) are highly genetically correlated, tea-drinking and height show no genetic correlation.


